src/main/java/com/epam/aidial/core/upstream/UpstreamRoute.java

```diff
@@ -36,41 +57,65 @@ public int retries() {
         return Math.min(retries, maxRetries);
     }
 
-    @Override
-    public boolean hasNext() {
-        return next < upstreams.size() && retries <= maxRetries;
+    /**
+     * Checks if upstream present (not null) and retry count does not exceed max retry attempts
+     *
+     * @return true if upstream available, false otherwise
+     */
+    public boolean available() {
+        return upstreamState != null && retries <= maxRetries;
```

Comment:
retries are monotonically grown and after some time upstream becomes unavailable due to the condition: `retries <= maxRetries` is false